src/csharp/Grpc.Core/Internal/AsyncCallBase.cs

```diff
@@ -0,0 +1,407 @@
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#endregion
+
+using System;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Grpc.Core.Internal;
+using Grpc.Core.Utils;
+
+namespace Grpc.Core.Internal
+{
+    /// <summary>
+    /// Base for handling both client side and server side calls.
+    /// Handles native call lifecycle and provides convenience methods.
+    /// </summary>
+    internal abstract class AsyncCallBase<TWrite, TRead>
+    {
+        readonly Func<TWrite, byte[]> serializer;
+        readonly Func<byte[], TRead> deserializer;
+
+        protected readonly CompletionCallbackDelegate sendFinishedHandler;
+        protected readonly CompletionCallbackDelegate readFinishedHandler;
+        protected readonly CompletionCallbackDelegate halfclosedHandler;
+
+        protected readonly object myLock = new object();
+
+        protected GCHandle gchandle;
+        protected CallSafeHandle call;
+        protected bool disposed;
+
+        protected bool started;
+        protected bool errorOccured;
+        protected bool cancelRequested;
+
+        protected AsyncCompletionDelegate sendCompletionDelegate;  // Completion of a pending send or sendclose if not null.
+        protected bool readPending;  // True if there is a read in progress.
+        protected bool readingDone;
+        protected bool halfcloseRequested;
+        protected bool halfclosed;
+        protected bool finished;  // True if close has been received from the peer.
+
+        // Streaming reads will be delivered to this observer. For a call that only does unary read it may remain null.
+        protected IObserver<TRead> readObserver;
+
+        public AsyncCallBase(Func<TWrite, byte[]> serializer, Func<byte[], TRead> deserializer)
+        {
+            this.serializer = Preconditions.CheckNotNull(serializer);
+            this.deserializer = Preconditions.CheckNotNull(deserializer);
+  
+            this.sendFinishedHandler = CreateBatchCompletionCallback(HandleSendFinished);
+            this.readFinishedHandler = CreateBatchCompletionCallback(HandleReadFinished);
+            this.halfclosedHandler = CreateBatchCompletionCallback(HandleHalfclosed);
+        }
+
+        /// <summary>
+        /// Requests cancelling the call.
+        /// </summary>
+        public void Cancel()
+        {
+            lock (myLock)
```

Comment:
Yes, it might not be. grpc_call_cancel should be threadsafe, but I am basically locking here to make sure that I can check the state of this class consistently.